draft-rfcs/const-evaluatable-checked.md

```diff
@@ -0,0 +1,18 @@
+# const evaluatable checked
+
+TODO: All of this :)
+
+## Requirements
+
+For this to work, const operations have to be deterministic and
+must not depend on any external state,
+at least when they are used in the type system.
+
+Using floats during CTFE is fully determinstic. So using
+them inside of the type system is fine. CTFE can however
+produce different results than what would happen on real hardware,
+but this is not a concern for const generics.
+
+Other sources of non-determinism are allocations. This non-determinism
+must however not be observed during const-evaluation (TODO: link to const-eval).
+Any references used in a constant are considered equal if their targets are equal, which is also determistic. (ref [val-trees](https://github.com/rust-lang/rust/issues/72396))
```

Comment:
Yes that works. :) It might be worth pointing out that raw pointers and function pointers must *not* be allowed, since they cannot be compared this way.

Reply:
yeah, the giant mess which is structural match :sparkles:

Reply:
luckily it's not too bad for const generics cause we don't have to care about structural match for values as we only look at it for types.

iirc dealing with consts in patterns is the actual hard part ^^

Reply:
> yeah, the giant mess which is structural match sparkles

I wasn't even talking about that.^^ But I hope valtrees can help fix that, too.

> luckily it's not too bad for const generics cause we don't have to care about structural match for values as we only look at types.

I don't follow. You care about when two values of some type `T` are equal. You care to do that deterministically. So you should ensure `T` does not involve raw pointers, function pointers, or trait objects.

Reply:
the way we forbid `&dyn Trait` and raw pointers is by checking that the types implement structural match.

> I don't follow. You care about when two values of some type `T` are equal. You care to do that deterministically. So you should ensure `T` does not involve raw pointers, function pointers, or trait objects.

As far as I know `None::<NonStructuralMatchTy>` is an allowed as a pattern. So for patterns we also use control flow analysis to check if a value can be used. At least that's what I infer from https://github.com/rust-lang/rust/pull/67343 and https://github.com/rust-lang/rust/issues/73448, didn't interact with that for a while.

This does not matter for const generics, as even if enum variants which are structurally match exist for a type, that type can still not be used as a const parameter type.

Reply:
Oh, that's what you mean. Sure, but the valtree framework can handle `None::<T>` for any `T` -- what really matters is if you can construct a valtree for the given *value*, i.e., it matters whether the given *value* contains a raw pointer, function pointer, or trait object. The type-based analysis just overapproximates that property.